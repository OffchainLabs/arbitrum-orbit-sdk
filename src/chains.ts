import { defineChain } from 'viem';
import {
  mainnet,
  holesky,
  arbitrum as arbitrumOne,
  arbitrumNova,
  sepolia,
  arbitrumSepolia,
} from 'viem/chains';

const nitroTestnodeL1 = defineChain({
  id: 1_337,
  network: 'Nitro Testnode L1',
  name: 'nitro-testnode-l1',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8545'],
    },
    public: {
      http: ['http://127.0.0.1:8545'],
    },
  },
  testnet: true,
});

const nitroTestnodeL2 = defineChain({
  id: 412_346,
  network: 'Nitro Testnode L2',
  name: 'nitro-testnode-l2',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8547'],
    },
    public: {
      http: ['http://127.0.0.1:8547'],
    },
  },
  testnet: true,
});

// the L3 deployed as part of the nitro-testnode framework uses a custom fee token
const nitroTestnodeL3 = defineChain({
  id: 333333,
  network: 'Nitro Testnode L3',
  name: 'nitro-testnode-l3',
  nativeCurrency: { name: 'AppTestToken', symbol: 'APP', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:3347'],
    },
    public: {
      http: ['http://127.0.0.1:3347'],
    },
  },
  testnet: true,
});

export const chains = [
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
] as const;

export {
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
};
