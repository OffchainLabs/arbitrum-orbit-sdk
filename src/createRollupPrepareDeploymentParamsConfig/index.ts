import { Chain, Client, Transport } from 'viem';

import {
  createRollupPrepareDeploymentParamsConfig as createRollupPrepareDeploymentParamsConfigV2Dot1,
  CreateRollupPrepareDeploymentParamsConfigParams as CreateRollupPrepareDeploymentParamsConfigParamsV2Dot1,
  CreateRollupPrepareDeploymentParamsConfigResult as CreateRollupPrepareDeploymentParamsConfigResultV2Dot1,
} from './v2.1';

import {
  createRollupPrepareDeploymentParamsConfig as createRollupPrepareDeploymentParamsConfigV3Dot1,
  CreateRollupPrepareDeploymentParamsConfigParams as CreateRollupPrepareDeploymentParamsConfigParamsV3Dot1,
  CreateRollupPrepareDeploymentParamsConfigResult as CreateRollupPrepareDeploymentParamsConfigResultV3Dot1,
} from './v3.1';

// Re-export the latest version types as defaults
export type CreateRollupPrepareDeploymentParamsConfigParams =
  CreateRollupPrepareDeploymentParamsConfigParamsV3Dot1;
export type CreateRollupPrepareDeploymentParamsConfigResult =
  CreateRollupPrepareDeploymentParamsConfigResultV3Dot1;

// Discriminated union function signature with conditional types
export function createRollupPrepareDeploymentParamsConfig<
  TChain extends Chain | undefined,
  TRollupCreatorVersion extends 'v2.1' | 'v3.1' | undefined = 'v3.1',
>(
  client: Client<Transport, TChain>,
  params: TRollupCreatorVersion extends 'v2.1'
    ? CreateRollupPrepareDeploymentParamsConfigParamsV2Dot1
    : CreateRollupPrepareDeploymentParamsConfigParamsV3Dot1,
  rollupCreatorVersion?: TRollupCreatorVersion,
): TRollupCreatorVersion extends 'v2.1'
  ? CreateRollupPrepareDeploymentParamsConfigResultV2Dot1
  : CreateRollupPrepareDeploymentParamsConfigResultV3Dot1 {
  const version = rollupCreatorVersion ?? 'v3.1';

  if (version === 'v2.1') {
    return createRollupPrepareDeploymentParamsConfigV2Dot1(
      client,
      params as CreateRollupPrepareDeploymentParamsConfigParamsV2Dot1,
    ) as any;
  }

  return createRollupPrepareDeploymentParamsConfigV3Dot1(
    client,
    params as CreateRollupPrepareDeploymentParamsConfigParamsV3Dot1,
  ) as any;
}
