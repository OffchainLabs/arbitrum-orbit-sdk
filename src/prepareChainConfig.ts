import { ExcludeSome, Prettify, RequireSome } from './types/utils';
import { ChainConfig, ChainConfigArbitrumParams } from './types/ChainConfig';

export const defaults = {
  homesteadBlock: 0,
  daoForkBlock: null,
  daoForkSupport: true,
  eip150Block: 0,
  eip150Hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
  eip155Block: 0,
  eip158Block: 0,
  byzantiumBlock: 0,
  constantinopleBlock: 0,
  petersburgBlock: 0,
  istanbulBlock: 0,
  muirGlacierBlock: 0,
  berlinBlock: 0,
  londonBlock: 0,
  clique: {
    period: 0,
    epoch: 0,
  },
  arbitrum: {
    EnableArbOS: true,
    AllowDebugPrecompiles: false,
    DataAvailabilityCommittee: false,
    InitialArbOSVersion: 32,
    GenesisBlockNum: 0,
    MaxCodeSize: 24_576,
    MaxInitCodeSize: 49_152,
  },
};

export type PrepareChainConfigParams = Prettify<
  Pick<ChainConfig, 'chainId'> & {
    arbitrum: PrepareChainConfigArbitrumParams;
  }
>;

export type PrepareChainConfigArbitrumParams = RequireSome<
  // exclude some fields that shouldn't be changed
  ExcludeSome<
    ChainConfigArbitrumParams,
    'EnableArbOS' | 'GenesisBlockNum' | 'AllowDebugPrecompiles'
  >,
  // make InitialChainOwner required
  'InitialChainOwner'
>;

export function prepareChainConfig(params: PrepareChainConfigParams): ChainConfig {
  return {
    ...defaults,
    chainId: params.chainId,
    arbitrum: { ...defaults.arbitrum, ...params.arbitrum },
  };
}
