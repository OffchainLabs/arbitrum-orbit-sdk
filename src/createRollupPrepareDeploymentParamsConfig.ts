import { Chain, Client, Transport } from 'viem';

import { ChainConfig } from './types/ChainConfig';
import { validateParentChain } from './types/ParentChain';
import { Prettify } from './types/utils';

import { createRollup } from './createRollup';
import { CreateRollupFunctionInputs } from './types/createRollupTypes';
import { prepareChainConfig } from './prepareChainConfig';

import { defaults } from './createRollupPrepareDeploymentParamsConfigDefaults';
import { getDefaultConfirmPeriodBlocks } from './getDefaultConfirmPeriodBlocks';
import { getDefaultSequencerInboxMaxTimeVariation } from './getDefaultSequencerInboxMaxTimeVariation';

export type CreateRollupPrepareDeploymentParamsConfigResult =
  CreateRollupFunctionInputs[0]['config'];

type RequiredKeys = 'chainId' | 'owner';
type RequiredParams = Pick<CreateRollupPrepareDeploymentParamsConfigResult, RequiredKeys>;
type OptionalParams = Partial<
  Omit<CreateRollupPrepareDeploymentParamsConfigResult, 'chainConfig' | RequiredKeys>
>;

export type CreateRollupPrepareDeploymentParamsConfigParams = Prettify<
  RequiredParams & { chainConfig?: ChainConfig } & OptionalParams
>;

/**
 * Creates the configuration object to be used with {@link createRollup}.
 *
 * @param {PublicClient} client Parent chain client
 * @param {Object} params Chain configuration parameters
 * @see https://docs.arbitrum.io/launch-orbit-chain/how-tos/customize-deployment-configuration
 * @see https://docs.arbitrum.io/launch-orbit-chain/reference/additional-configuration-parameters
 * @param {string} params.owner
 * @param {BigInt} params.chainId
 * @param {ChainConfig} [params.chainConfig]
 * @param {BigInt} [params.confirmPeriodBlocks]
 * @param {BigInt} [params.extraChallengeTimeBlocks]
 * @param {string} [params.stakeToken]
 * @param {BigInt} [params.baseStake]
 * @param {string} [params.wasmModuleRoot]
 * @param {string} [params.loserStakeEscrow]
 * @param {BigInt} [params.genesisBlockNum]
 * @param {Object} [params.sequencerInboxMaxTimeVariation]
 * @param {BigInt} [params.sequencerInboxMaxTimeVariation.delayBlocks]
 * @param {BigInt} [params.sequencerInboxMaxTimeVariation.futureBlocks]
 * @param {BigInt} [params.sequencerInboxMaxTimeVariation.delaySeconds]
 * @param {BigInt} [params.sequencerInboxMaxTimeVariation.futureSeconds]
 *
 * @returns {Object} {@link CreateRollupPrepareDeploymentParamsConfigResult}
 *
 * @example
 * const config = createRollupPrepareDeploymentParamsConfig(parentPublicClient, {
 *   chainId: BigInt(chainId),
 *   owner: deployer.address,
 *   chainConfig: prepareChainConfig({
 *     chainId,
 *     arbitrum: {
 *       InitialChainOwner: deployer.address,
 *       DataAvailabilityCommittee: true,
 *     },
 *   }),
 * });
 */
export function createRollupPrepareDeploymentParamsConfig<
  TTransport extends Transport = Transport,
  TChain extends Chain | undefined = Chain | undefined,
>(
  client: Client<TTransport, TChain>,
  { chainConfig, ...params }: CreateRollupPrepareDeploymentParamsConfigParams,
): CreateRollupPrepareDeploymentParamsConfigResult {
  const parentChainId = validateParentChain(client);

  const defaultsBasedOnParentChain = {
    confirmPeriodBlocks: getDefaultConfirmPeriodBlocks(parentChainId),
    sequencerInboxMaxTimeVariation: getDefaultSequencerInboxMaxTimeVariation(parentChainId),
  };

  return {
    ...defaults,
    ...defaultsBasedOnParentChain,
    ...params,
    chainConfig: JSON.stringify(
      chainConfig ??
        prepareChainConfig({
          chainId: Number(params.chainId),
          arbitrum: { InitialChainOwner: params.owner },
        }),
    ),
  };
}
