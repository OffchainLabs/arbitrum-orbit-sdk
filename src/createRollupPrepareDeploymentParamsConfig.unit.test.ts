import { it, expect } from 'vitest';
import { Address, createPublicClient, http } from 'viem';

import { arbitrumOne, arbitrumSepolia } from './chains';
import { prepareChainConfig } from './prepareChainConfig';
import { createRollupPrepareDeploymentParamsConfig } from './createRollupPrepareDeploymentParamsConfig';

const chainId = 69_420n;
const vitalik: `0x${string}` = '0xd8da6bf26964af9d7eed9e03e53415d37aa96045';

function getOverrides({ owner, chainId }: { owner: Address; chainId: bigint }) {
  return {
    owner,
    chainId,
    chainConfig: prepareChainConfig({
      chainId: Number(chainId),
      arbitrum: {
        InitialChainOwner: owner,
        InitialArbOSVersion: 30,
        DataAvailabilityCommittee: true,
      },
    }),
    confirmPeriodBlocks: 4200n,
    extraChallengeTimeBlocks: 5n,
    loserStakeEscrow: '0x0000000000000000000000000000000000000001',
    sequencerInboxMaxTimeVariation: {
      delayBlocks: 200n,
      delaySeconds: 5n,
      futureBlocks: 100n,
      futureSeconds: 1n,
    },
    stakeToken: '0x0000000000000000000000000000000000000002',
    wasmModuleRoot: '0xWasmModuleRoot',
  } as const;
}

it('creates config for a chain on top of arbitrum one with defaults', () => {
  const arbitrumOneClient = createPublicClient({
    chain: arbitrumOne,
    transport: http(),
  });

  expect(
    createRollupPrepareDeploymentParamsConfig(arbitrumOneClient, {
      owner: vitalik,
      chainId,
    }),
  ).toMatchSnapshot();
});

it('creates config for a chain on top of arbitrum one with overrides', () => {
  const arbitrumOneClient = createPublicClient({
    chain: arbitrumOne,
    transport: http(),
  });

  expect(
    createRollupPrepareDeploymentParamsConfig(
      arbitrumOneClient,
      getOverrides({ owner: vitalik, chainId }),
    ),
  ).toMatchSnapshot();
});

it('creates config for a chain on top of arbitrum sepolia with defaults', () => {
  const arbitrumSepoliaClient = createPublicClient({
    chain: arbitrumSepolia,
    transport: http(),
  });

  expect(
    createRollupPrepareDeploymentParamsConfig(arbitrumSepoliaClient, {
      owner: vitalik,
      chainId,
    }),
  ).toMatchSnapshot();
});

it('creates config for a chain on top of arbitrum sepolia with overrides', () => {
  const arbitrumSepoliaClient = createPublicClient({
    chain: arbitrumSepolia,
    transport: http(),
  });

  expect(
    createRollupPrepareDeploymentParamsConfig(
      arbitrumSepoliaClient,
      getOverrides({ owner: vitalik, chainId }),
    ),
  ).toMatchSnapshot();
});
