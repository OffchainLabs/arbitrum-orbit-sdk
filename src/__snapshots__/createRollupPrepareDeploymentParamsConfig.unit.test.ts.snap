// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`creates config for a chain on top of arbitrum one with defaults 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":32,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 45818n,
  "extraChallengeTimeBlocks": 0n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000000",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 28800n,
    "delaySeconds": 345600n,
    "futureBlocks": 300n,
    "futureSeconds": 3600n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000000",
  "wasmModuleRoot": "0x184884e1eb9fefdc158f6c8ac912bb183bf3cf83f0090317e0bc4ac5860baa39",
}
`;

exports[`creates config for a chain on top of arbitrum one with overrides 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":true,\\"InitialArbOSVersion\\":20,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 4200n,
  "extraChallengeTimeBlocks": 5n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000001",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 200n,
    "delaySeconds": 5n,
    "futureBlocks": 100n,
    "futureSeconds": 1n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000002",
  "wasmModuleRoot": "0xWasmModuleRoot",
}
`;

exports[`creates config for a chain on top of arbitrum sepolia with defaults 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":32,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 150n,
  "extraChallengeTimeBlocks": 0n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000000",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 28800n,
    "delaySeconds": 345600n,
    "futureBlocks": 300n,
    "futureSeconds": 3600n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000000",
  "wasmModuleRoot": "0x184884e1eb9fefdc158f6c8ac912bb183bf3cf83f0090317e0bc4ac5860baa39",
}
`;

exports[`creates config for a chain on top of arbitrum sepolia with overrides 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":true,\\"InitialArbOSVersion\\":20,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 4200n,
  "extraChallengeTimeBlocks": 5n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000001",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 200n,
    "delaySeconds": 5n,
    "futureBlocks": 100n,
    "futureSeconds": 1n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000002",
  "wasmModuleRoot": "0xWasmModuleRoot",
}
`;

exports[`creates config for a chain on top of base sepolia with defaults 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":32,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 900n,
  "extraChallengeTimeBlocks": 0n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000000",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 172800n,
    "delaySeconds": 345600n,
    "futureBlocks": 1800n,
    "futureSeconds": 3600n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000000",
  "wasmModuleRoot": "0x184884e1eb9fefdc158f6c8ac912bb183bf3cf83f0090317e0bc4ac5860baa39",
}
`;

exports[`creates config for a chain on top of base with defaults 1`] = `
{
  "baseStake": 100000000000000000n,
  "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":false,\\"InitialArbOSVersion\\":32,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0xd8da6bf26964af9d7eed9e03e53415d37aa96045\\"},\\"chainId\\":69420}",
  "chainId": 69420n,
  "confirmPeriodBlocks": 274908n,
  "extraChallengeTimeBlocks": 0n,
  "genesisBlockNum": 0n,
  "loserStakeEscrow": "0x0000000000000000000000000000000000000000",
  "owner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
  "sequencerInboxMaxTimeVariation": {
    "delayBlocks": 172800n,
    "delaySeconds": 345600n,
    "futureBlocks": 1800n,
    "futureSeconds": 3600n,
  },
  "stakeToken": "0x0000000000000000000000000000000000000000",
  "wasmModuleRoot": "0x184884e1eb9fefdc158f6c8ac912bb183bf3cf83f0090317e0bc4ac5860baa39",
}
`;
